File: Day31/bitwise-ors-of-subarrays.js
// Time: O(N * LogW)  ::  Space: O(N)

var subarrayBitwiseORs = function(arr) {
    let ans = new Set();
    let cur = new Set([0]);
    for (let x of arr) {
        let temp = new Set();
        temp.add(x);
        for (let y of cur) {
            temp.add(x | y);
        }
        cur = temp;
        for (let val of cur) {
            ans.add(val);
        }
    }
    return ans.size;
};

File: Day31/bitwise-ors-of-subarrays.java
// Time: O(N * LogW)  ::  Space: O(N)

import java.util.*;

class Solution {
    public int subarrayBitwiseORs(int[] arr) {
        Set<Integer> ans = new HashSet<>();
        Set<Integer> cur = new HashSet<>();
        cur.add(0);
        for (int x : arr) {
            Set<Integer> temp = new HashSet<>();
            temp.add(x);
            for (int y : cur) {
                temp.add(x | y);
            }
            cur = temp;
            ans.addAll(cur);
        }
        return ans.size();
    }
}
